File: code/apps/web/src/lib/api/optimistic.ts
/**
 * Optimistic UI, confined to where it is safe (SRS-WEB-007).
 *
 * The verification clause draws the line: "clinical/financial finalization is
 * server-confirmed before UI success". Optimistic UI makes an application feel
 * fast by showing the result before the server agrees, and in most software
 * the cost of being wrong is a flicker. In a hospital it is a clinician
 * reading "administered" for a dose that was never recorded, or a cashier
 * seeing a payment that did not settle.
 *
 * So this module does not offer a general optimistic wrapper. It offers one
 * for operations that are explicitly declared reversible, and it refuses at
 * runtime to apply optimism to anything marked as a finalization — because a
 * rule that lives only in a code review is a rule that lasts until the first
 * deadline.
 */

/** What kind of operation is being performed. */
export type OperationKind =
	/**
	 * Reversible and non-clinical: renaming a saved view, toggling a column,
	 * marking a task read. Rolling back is invisible to anyone but the user
	 * who did it.
	 */
	| 'reversible'
	/**
	 * Clinical or financial finalization: signing a note, administering a
	 * medication, taking a payment, issuing an invoice. Never optimistic.
	 */
	| 'finalization';

/** Thrown when optimism is applied where it must not be. */
export class OptimismForbiddenError extends Error {
	constructor(readonly operation: string) {
		super(
			`${operation} is a finalization: the UI must not show success before the ` +
				`server confirms it (SRS-WEB-007)`
		);
		this.name = 'OptimismForbiddenError';
	}
}

/** The visible state of an optimistic operation. */
export type OptimisticState = 'idle' | 'pending' | 'confirmed' | 'rolled-back';

/** What the caller sees while and after an operation runs. */
export interface OptimisticResult<T> {
	readonly state: OptimisticState;
	readonly value: T;
	/** Set when the operation failed and the value was rolled back. */
	readonly error: Error | null;
}

/**
 * Applies a change optimistically and rolls it back if the server refuses.
 *
 * kind is required and checked. There is no default, because a default would
 * be 'reversible' — the convenient one — and the first finalization written by
 * someone in a hurry would inherit it.
 *
 * onChange is called at every transition rather than only at the end, so the
 * rollback is visible: SRS-WEB-007 asks for optimism only where "rollback is
 * visible", and a rollback the user never sees is indistinguishable from the
 * operation having worked.
 */
export async function optimistic<T>(params: {
	readonly operation: string;
	readonly kind: OperationKind;
	readonly current: T;
	readonly optimistic: T;
	readonly commit: () => Promise<T>;
	readonly onChange: (result: OptimisticResult<T>) => void;
}): Promise<OptimisticResult<T>> {
	if (params.kind === 'finalization') {
		throw new OptimismForbiddenError(params.operation);
	}

	params.onChange({ state: 'pending', value: params.optimistic, error: null });

	try {
		const confirmed = await params.commit();
		const result: OptimisticResult<T> = { state: 'confirmed', value: confirmed, error: null };
		params.onChange(result);
		return result;
	} catch (cause) {
		// Rolled back to the value from before the attempt, not to the
		// optimistic one. Leaving the optimistic value on screen after a
		// failure is the specific harm this module exists to prevent.
		const result: OptimisticResult<T> = {
			state: 'rolled-back',
			value: params.current,
			error: cause instanceof Error ? cause : new Error(String(cause))
		};
		params.onChange(result);
		return result;
	}
}

/**
 * Runs a finalization, showing success only once the server has confirmed.
 *
 * Deliberately has no optimistic parameter. The pending state exists so the UI
 * can disable the button and show progress — what it must not do is show the
 * outcome.
 */
export async function confirmed<T>(params: {
	readonly operation: string;
	readonly commit: () => Promise<T>;
	readonly onChange: (result: OptimisticResult<T | null>) => void;
}): Promise<OptimisticResult<T | null>> {
	params.onChange({ state: 'pending', value: null, error: null });

	try {
		const value = await params.commit();
		const result: OptimisticResult<T | null> = { state: 'confirmed', value, error: null };
		params.onChange(result);
		return result;
	} catch (cause) {
		const result: OptimisticResult<T | null> = {
			state: 'rolled-back',
			value: null,
			error: cause instanceof Error ? cause : new Error(String(cause))
		};
		params.onChange(result);
		return result;
	}
}

/**
 * Reports whether an operation may be shown optimistically.
 *
 * Exported so a component can decide which helper to call without duplicating
 * the rule, and so a test can assert the classification of a specific
 * operation rather than only the behaviour of the wrapper.
 */
export function mayBeOptimistic(kind: OperationKind): boolean {
	return kind === 'reversible';
}

File: code/apps/mobile/lib/src/api/optimistic.dart
/// Optimistic updates, and where they are forbidden (SRS-WEB-007).
///
/// The requirement permits optimistic UI "only where server conflict semantics
/// are defined and rollback is visible". Both halves are constraints, and the
/// second is the one a mobile app gets wrong: a phone is frequently offline, so
/// the gap between showing a result and the server confirming it can be
/// minutes rather than milliseconds, and a rollback that happens quietly while
/// the screen is in a pocket is a change the user never learns about.
///
/// So an optimistic result here is never silently reverted. A rollback carries
/// the reason and stays visible until it is acknowledged, and the operations
/// where that is not good enough are refused outright rather than left to a
/// caller's judgement.
library;

import 'dart:async';

/// The kinds of operation an optimistic update may be attempted for.
enum OperationKind {
  /// A user's own view preference. Wrong for nobody but them, and instantly
  /// re-settable.
  preference,

  /// Marking something read or acknowledged in a worklist.
  acknowledgement,

  /// Anything that writes to the clinical record.
  clinicalWrite,

  /// Anything that moves money.
  financialWrite,

  /// Anything that administers or prescribes a medicine.
  medicationWrite,
}

/// Thrown when optimism is attempted where it is not permitted.
class OptimismForbiddenError implements Exception {
  OptimismForbiddenError(this.kind);

  final OperationKind kind;

  @override
  String toString() =>
      'OptimismForbiddenError: $kind must not be shown as done before the '
      'server confirms it';
}

/// Whether an optimistic update is permitted for this kind of operation.
///
/// Clinical, financial and medication writes are refused. The reason is not
/// that they are important in the abstract — it is that showing them as done
/// changes what the user does next: a nurse who sees a dose recorded does not
/// record it again, and a rollback arriving after they have walked away leaves
/// a dose nobody gave and nobody knows is missing.
bool mayBeOptimistic(OperationKind kind) =>
    kind == OperationKind.preference || kind == OperationKind.acknowledgement;

/// Where an optimistic value stands.
enum OptimisticState { idle, pending, confirmed, rolledBack }

/// An optimistic value and what became of it.
class OptimisticResult<T> {
  const OptimisticResult({
    required this.value,
    required this.state,
    this.reason = '',
  });

  /// What the screen should show: the optimistic value while pending, the
  /// server's value once confirmed, the original once rolled back.
  final T value;

  final OptimisticState state;

  /// Why it was rolled back. Non-empty only for [OptimisticState.rolledBack],
  /// and shown to the user rather than logged — a silent revert is a change
  /// the user never learns about.
  final String reason;

  bool get needsAcknowledgement => state == OptimisticState.rolledBack;
}

/// Applies an update optimistically, rolling back visibly on failure.
///
/// [optimistic] is shown immediately; [commit] is the server call. On failure
/// the original value comes back with the reason attached, so the caller
/// renders the revert rather than quietly dropping it.
Future<OptimisticResult<T>> optimistic<T>({
  required OperationKind kind,
  required T original,
  required T optimistic,
  required Future<T> Function() commit,
  void Function(OptimisticResult<T>)? onChange,
}) async {
  if (!mayBeOptimistic(kind)) {
    throw OptimismForbiddenError(kind);
  }

  onChange?.call(
    OptimisticResult(value: optimistic, state: OptimisticState.pending),
  );

  try {
    final confirmed = await commit();
    final result =
        OptimisticResult(value: confirmed, state: OptimisticState.confirmed);
    onChange?.call(result);
    return result;
  } catch (error) {
    final result = OptimisticResult(
      value: original,
      state: OptimisticState.rolledBack,
      // The message, not the exception: it is going on a screen.
      reason: _describe(error),
    );
    onChange?.call(result);
    return result;
  }
}

/// Runs an operation that must not be shown as done until the server says so.
///
/// The counterpart to [optimistic], and the one every clinical path uses. The
/// screen shows pending, then the server's answer — never a guess.
Future<OptimisticResult<T>> confirmed<T>({
  required T original,
  required Future<T> Function() commit,
  void Function(OptimisticResult<T>)? onChange,
}) async {
  onChange?.call(
    OptimisticResult(value: original, state: OptimisticState.pending),
  );
  try {
    final settled = await commit();
    final result =
        OptimisticResult(value: settled, state: OptimisticState.confirmed);
    onChange?.call(result);
    return result;
  } catch (error) {
    final result = OptimisticResult(
      value: original,
      state: OptimisticState.rolledBack,
      reason: _describe(error),
    );
    onChange?.call(result);
    return result;
  }
}

String _describe(Object error) {
  final text = error.toString();
  return text.isEmpty ? 'The change could not be saved.' : text;
}
